### PR TITLE
feat(2195): mini-pulse-history component

### LINKED ISSUE #2195: Mini pulse history component
We are looking for this kind of component to be added to the series and spotlight.

This is blocker #1680 probably?
The component should feed data of recent buy events.

In the series, show buys for the last 30 days 
In the spotlight of particular authors sales for the last 30 days

image

### Ref
- https://steamdb.info/ ht @vikiival 
- https://github.com/kodadot
- https://github.com/unjs
- #1680

### PR DESCRIPTION
**Pull request by KngZhi**
**Thank you for your contribution** to the [KodaDot NFT gallery](https://kodadot.xyz).
👇 \_ Let's make a quick check before the contribution.

### PR type

- [ ] Bugfix
- [x] Feature
- [ ] Refactoring

### What's new?

- [x] PR related #2195
- [ ] <brief_description_of_what_I've_added>

### Before submitting Pull Request, please make sure:

- [x] My contribution builds **clean without any errors or warnings**
- [ ] I've merged recent default branch -- **main** and I've no conflicts
- [x] I've tried to respect high code quality standards
- [x] I've didn't break any original functionality
- [x] I've posted a screenshot of demonstrated change in this PR

### Optional

- [ ] I've tested it at </rmrk/collection/26902bc2f7c20c546a-1FVG7>
- [ ] I've tested PR on mobile and everything seems works
- [ ] I found edge cases
- [ ] I've written some unit tests 🧪

### Had issue bounty label?

- [x] Fill up your KSM address: [Payout](https://beta.kodadot.xyz/transfer?target=DaWoVNNhBTdqJ5iPN3UcaLeamkPpJeGWdrs8LhvX9TQbadc)

### Community participation

- [x] [Are you at KodaDot Discord?](https://discord.gg/35hzy2dXXh)

### Screenshot

- [x] My fix has changed **something** on UI; a screenshot is best to understand changes for others.

image

### CODE REVIEW

**Comment by KngZhi** [comment]
Hey @yangwao this is a demo page, do you prefer this UI for desktop and mobile.

Besides, I think we need to update the resolver to retrieve relevant data, but I'm kinda stuck with that, is there anyone I can resort to or communicate with.

**Comment by yangwao** [comment]
nice!
 
image

**Comment by KngZhi** [comment]
Brief Summary:

feat:
1. a mini-history component for the series and spotlight(EChart)
2. queries for historical data, and PLEASE correct me if there had any more efficient or clean way to do GraphQL. 😄 [1]

fix:
1. using computed data for the spotlight
3. [spotlight] fix pageSize problem
4. align-center for each column

refactor:
1. using computed data as the datasource for the spotlight

Following TODOs(for other PRs):
1. Since the EChart supports SVG format, which is more friendly to Mobile, I want to refactor other pages that use Chart.js
1. Clean useless code inside spotlight and series, besides, refactor some duplicate code.

**Review by vikiival — CHANGES_REQUESTED** [review]
This should be done via rubick resolver

File: package.json [hunk 1]
```diff
@@ -81,6 +81,7 @@
     "css-loader": "5.2.7",
     "cypress": "^9.5.4",
     "date-fns": "^2.28.0",
+    "echarts": "^5.3.2",
```
[the remark this answers is not in the document]
Comment (KngZhi) [reply package.json:84]: Echarts are not welcomed? 🤔

File: components/series/utils.ts [hunk 2]
```diff
@@ -82,10 +82,39 @@ const onlyEvents = (nft: SimpleSeriesNFT) => nft.events
 export const toSort = (sortBy: SortType): string =>
   `${sortBy.field}_${sortBy.value}`
 
-const today = startOfToday()
+export const today = new Date()
```
[the remark this answers is not in the document]
Comment (KngZhi) [reply components/series/utils.ts:85]: The original would before one day.

**Comment by KngZhi** [comment]
Besides, could you check another PR #54 on rubick.

File: package.json (hunk 1 above)
[the remark this answers is not in the document]
Comment (vikiival) [reply package.json:84]: We had `Echarts` in the past and we have seen some performance issues

[the remark this answers is not in the document]
Comment (yangwao) [reply package.json:84]: Can we use the existing charting engine we are using for collection activity ?

[the remark this answers is not in the document]
Comment (KngZhi) [reply package.json:84]: Okay, I will check with `Chart.js`.

File: components/spotlight/SpotlightTable.vue [hunk 3]
```diff
@@ -250,19 +309,76 @@ export default class SpotlightTable extends mixins(
...
+      }, {})
+    this.data.forEach((row) => {
+      if (solds[row.id]) {
+        this.$set(row, 'soldHistory', axisLize(solds[row.id]))
+        // row.soldHistory = axisLize(solds[row.id])
+      }
+    })
+
+    this.isLoading = false
+  }
```
[the remark this answers is not in the document]
Comment (KngZhi) [reply components/spotlight/SpotlightTable.vue:354]: hold when [Rubick PR](https://github.com/kodadot/rubick/issues/58) merged.

File: package.json (hunk 1 above)
[the remark this answers is not in the document]
Comment (KngZhi) [reply package.json]: Fixed with `Chart.js`

**Comment by yangwao** [comment]
> Besides, could you check another PR #54 on rubick.

you mean this one https://github.com/kodadot/rubick/pull/54 :)

**Comment by KngZhi** [comment]
> > Besides, could you check another PR #54 on rubick.
> 
> you mean this one [kodadot/rubick#54](https://github.com/kodadot/rubick/pull/54) :)

Haha, yes, the [issue#877](https://github.com/kodadot/nft-gallery/issues/877) can not continued if [kodadot/rubick#54](https://github.com/kodadot/rubick/pull/54) can not be merged.

File: components/series/SeriesTable.vue [hunk 4]
```diff
@@ -353,17 +365,71 @@ export default class SeriesTable extends mixins(PrefixMixin) {
...
+      .map((e) => ({
+        ...e.nft.collection,
+        timestamp: e.timestamp.replace(/(T.*?$)/g, ''),
+      }))
+      .reduce((res, e) => {
+        const { id, timestamp: ts } = e
+        if (!res[id]) {
+          res[id] = Object.assign({}, defaultBuyEvents)
+        }
+        res[id][ts] += 1
```
Comment (vikiival) [inline components/series/SeriesTable.vue:393]: ooof

File: components/series/SeriesTable.vue [hunk 5]
```diff
@@ -353,17 +365,71 @@ export default class SeriesTable extends mixins(PrefixMixin) {
...
+      xAxisList: Object.keys(obj),
+      yAxisList: Object.values(obj),
+    })
+
+    const ids = collectionEntities.map((c: Collection) => c.id)
+
+    const buyEvents = (await this.fetchCollectionEvents(ids))
+      .map((e) => ({
+        ...e.nft.collection,
+        timestamp: e.timestamp.replace(/(T.*?$)/g, ''),
```
Comment (vikiival) [inline components/series/SeriesTable.vue:386]: create date utils or use the IntlFormatter

File: components/shared/PulseChart.vue [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+  renderChart() {
+    const ctx = (
+      document?.getElementById(this.chartId) as HTMLCanvasElement
+    )?.getContext('2d')
+
+    const data = {
+      labels: this.labels,
+      datasets: [
+        {
+          label: 'My First Dataset',
```
Comment (vikiival) [inline components/shared/PulseChart.vue:52]: 🤔

**Review by vikiival — CHANGES_REQUESTED** [review]
Small things

File: components/series/SeriesTable.vue (hunk 4 above)
Comment (KngZhi) [reply components/series/SeriesTable.vue:393]: Do you mind expanding the review comment?

Some thoughts:

Currently, it's quite dirty for the part of initialization and code related to `this.fetchCollectionEvents`, maybe we can hold that when we finished the resolver, then the transformation of fetched data would be much neater or not needed anymore. I can come back to refactor that.

Comment (KngZhi) [reply components/series/SeriesTable.vue:393]: I am willing to work on that resolver as well

get a keen desire 😄 to write code about that.

**Comment by yangwao** [comment]
👀

**Comment by vikiival** [comment]
> 👀

https://github.com/kodadot/rubick/pull/62

**Comment by KngZhi** [comment]
Is that possible that we can merge this PR firstly, It hung here for a quite long time. I am worrying this would be outdated with `main` branch for a long distance. The review comment mainly focuses on the data process, I would refactor that once for all in the new resolver, it will be much cleaner to do that in graphQL.

would finish that ASAP, and then I would refactor that.

**Review by vikiival — APPROVED** [review]

**Comment by KngZhi** [comment]
@yangwao could you share some opinions, let us merge firstly since all front-end part is done, and I would refactor the back-end code ASAP. We open a new PR.

> @vikiival Is that possible that we can merge this PR firstly, It hung here for a quite long time. I am worrying this would be outdated with `main` branch for a long distance. The review comment mainly focuses on the data process, I would refactor that once for all in the new resolver, it will be much cleaner to do that in graphQL.
> 
> would finish that ASAP, and then I would refactor that.

**Comment by yangwao** [comment]
> Let us merge firstly since all front-end part is done, and I would refactor the back-end code ASAP. We open a new PR.

okay, then I would propose to resolve conflicts, merge this now, but don't close parent issue, do partial payout and then you can come with new PR?

**Comment by yangwao** [comment]
pay 150 usd

**Comment by yangwao** [comment]
Wait, some graphs seems suspicious same, why is like that?

https://user-images.githubusercontent.com/5887929/166924486-b3533350-bdb2-4b8a-bed8-9305edd2a0d5.mov

**Comment by yangwao** [comment]
Looks like overlay bug 

image

**Comment by KngZhi** [comment]
> Wait, some graphs seem suspicious same, why is it like that?
> 
>  Screen.Recording.2022-05-05.at.14.37.35.mov

hmm, seems some data duplicated. Will fix it right now.

**Comment by KngZhi** [comment]
@yangwao 
https://vimeo.com/706753042/0a68a270a5

**Comment by yangwao** [comment]
@KngZhi you can directly upload videos to github input box to save you time like on this video :)


https://user-images.githubusercontent.com/5887929/167088257-41a7a37a-8981-47a5-bb8d-060a2d47e757.mov

**Comment by yangwao** [comment]
pay 150 usd

**Comment by yangwao** [comment]
pay 150 usd

**Comment by yangwao** [comment]
> Error completing request

lol what payout bot 👀

**Comment by yangwao** [comment]
pay 100 usd

**Comment by KngZhi** [comment]
> @KngZhi you can directly upload videos to github input box to save you time like on this video :)
> 
>  Screen.Recording.2022-05-06.at.09.37.25.mov

@yangwao oh, my, thanks, learned a new technique, hahaha 😄 

https://user-images.githubusercontent.com/11361147/167088967-7c636918-c00c-4f14-a046-1dcf5d5d7ae8.mov

**Comment by yangwao** [comment]
pay 100 usd

**Comment by yangwao** [comment]
oh got it, your address has 47 chars when I matching only 46 hmm, got it

**Comment by KngZhi** [comment]
Okay.

**Comment by yangwao** [comment]
pay 100 usd

**Comment by vikiival** [comment]
🚀

**Comment by yangwao** [comment]
pay 100 usd

**Comment by yangwao** [comment]
pay 100 usd

**Comment by yangwao** [comment]
😍 Perfect, I’ve sent the payout 
💵 $100 @ 123.95 USD/KSM ~ 0.807 $KSM
🧗 [DaWoVNNhBTdqJ5iPN3UcaLeamkPpJeGWdrs8LhvX9TQbadc](https://kusama.subscan.io/account/DaWoVNNhBTdqJ5iPN3UcaLeamkPpJeGWdrs8LhvX9TQbadc)
🔗 [0x52329283d8a8ae1d8806f937f74f5260db1a9f3de2f23c94b82a90b1965db7da](https://kusama.subscan.io/extrinsic/0x52329283d8a8ae1d8806f937f74f5260db1a9f3de2f23c94b82a90b1965db7da)

🪅 Let’s grab another issue and get rewarded! 
🪄 [github.com/kodadot/nft-gallery/issues](https://github.com/kodadot/nft-gallery/issues/)

**Pull request merged**